hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java

```diff
@@ -0,0 +1,560 @@
+      return cacheDeletedKeySet;
+    }
+  }
+
+  /**
+   * Implement a min heap iterator to find the smaller
+   * lexicographically sorted string.
+   */
+  private static class MinHeapIterator implements
+      Iterator<HeapEntry<? extends WithParentObjectId>>, Closeable {
+
+    private final PriorityQueue<HeapEntry
+        <? extends WithParentObjectId>> minHeap = new PriorityQueue<>();
+    private final ArrayList<Iterator<HeapEntry
+        <? extends WithParentObjectId>>> iterators = new ArrayList<>();
+
+    private final RawIter<OmDirectoryInfo> rawDirIter;
+    private final RawIter<OmKeyInfo> rawFileIter;
+
+    private final CacheIter<OmKeyInfo> cacheFileIter;
+    private final CacheIter<OmDirectoryInfo> cacheDirIter;
+
+    MinHeapIterator(OMMetadataManager omMetadataManager, String prefixKey,
+                    BucketLayout bucketLayout, String startKey,
+                    String volumeName, String bucketName) throws IOException {
+
+      omMetadataManager.getLock().acquireReadLock(BUCKET_LOCK, volumeName,
+          bucketName);
+      cacheDirIter =
+          new CacheIter<>(EntryType.DIR_CACHE,
+              omMetadataManager.getDirectoryTable().cacheIterator(),
+              startKey, prefixKey);
+
+      cacheFileIter =
+          new CacheIter<>(EntryType.FILE_CACHE,
+              omMetadataManager.getKeyTable(bucketLayout).cacheIterator(),
+              startKey, prefixKey);
+
+      rawDirIter =
+          new RawIter<>(EntryType.RAW_DIR_DB,
+              omMetadataManager.getDirectoryTable().iterator(),
+              prefixKey, startKey, cacheDirIter.getDeletedKeySet());
+
+      rawFileIter =
+          new RawIter<>(EntryType.RAW_FILE_DB,
+              omMetadataManager.getKeyTable(bucketLayout).iterator(),
+              prefixKey, startKey, cacheFileIter.getDeletedKeySet());
+
+      omMetadataManager.getLock().releaseReadLock(BUCKET_LOCK, volumeName,
+          bucketName);
+
+      iterators.add(EntryType.DIR_CACHE.ordinal(), cacheDirIter);
+      iterators.add(EntryType.FILE_CACHE.ordinal(), cacheFileIter);
+      iterators.add(EntryType.RAW_FILE_DB.ordinal(), rawFileIter);
+      iterators.add(EntryType.RAW_DIR_DB.ordinal(), rawDirIter);
+      insertFirstElement();
+
+    }
+
+    public void insertFirstElement() {
+      for (Iterator<HeapEntry<? extends WithParentObjectId>> iter :
+          iterators) {
+        if (iter.hasNext()) {
+          minHeap.add(iter.next());
+        }
+      }
+    }
+
+    public boolean hasNext() {
+      return !minHeap.isEmpty();
+    }
+
+    public HeapEntry<? extends WithParentObjectId> next() {
+      HeapEntry<? extends WithParentObjectId> heapEntry = minHeap.remove();
+      Iterator<HeapEntry<? extends WithParentObjectId>> iter =
+          iterators.get(heapEntry.entryType.ordinal());
+      if (iter.hasNext()) {
+        minHeap.add(iter.next());
+      }
+
+      return heapEntry;
+    }
+
+    public void close() throws IOException {
+    }
```

Comment:
Please close all the iterators and which inturn should close `tableIterator.close();`